src/Utilities/MakeNaN.hpp

```diff
@@ -0,0 +1,40 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <limits>
+
+namespace detail {
+
+// This will need to be updated as needs arise for signaling NaNs in containers
+// apart from std::complex.
+template <typename T, typename = std::nullptr_t>
+struct MakeSignalingNaNImpl;
+
+// for fundamental types, the NaN is provided by std::numeric_limits
+template <typename T>
+struct MakeSignalingNaNImpl<
+    T, Requires<std::numeric_limits<T>::has_signaling_NaN>> {
+  const T value = std::numeric_limits<T>::signaling_NaN();
+};
+
+// for complex types, make both entries a NaN
+template <typename T>
+struct MakeSignalingNaNImpl<
+    std::complex<T>, Requires<std::numeric_limits<T>::has_signaling_NaN>> {
+  const std::complex<T> value =
+      std::complex<T>(std::numeric_limits<T>::signaling_NaN(),
+                      std::numeric_limits<T>::signaling_NaN());
+};
+
+}  // namespace detail
+
+/// \ingroup UtilitiesGroup
+
+/// \brief Returns an appropriate signaling NaN for fundamantal or multi-field
+/// types.
+template <typename T>
+T make_signaling_NaN() {
```

Comment:
Just make two overloaded functions and remove the requires. The compiler error will be reasonable enough and shouldn't be encountered often, if ever. I don't see a reason to pay for SFINAE and all the extra code in this case.
```cpp
template <typename T>
T make_signaling_NaN(const T& /*meta*/) noexcept {
  return std::numeric_limits<T>::signaling_NaN();
}

template <typename T>
std::complex<T> make_signaling_NaN(const std::complex<T>& /*meta*/) noexcept {
  return {std::numeric_limits<T>::signaling_NaN(), std::numeric_limits<T>::signaling_NaN());
}
```

Reply:
This seems fine, but I would prefer the type to be an explicit template argument rather than passed in the argument, as that acts more similarly to the `std::numeric_limits`.

Reply:
`numeric_limits` is a struct with static functions, this already doesn't behave like `numeric_limits`. You could have a generic:
```cpp
template <typename T>
T make_signaling_NaN() noexcept {
  return make_signaling_NaN(static_cast<T>(0));
}
```
I think that'll do what you want without all the overhead of having to do memoization and SFINAE on top of pattern matching.